src/main/java/calculator/StringCalculator.java

```diff
@@ -0,0 +1,36 @@
+package calculator;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class StringCalculator {
+
+    public int add(String text) throws RuntimeException {
+
+        if (text == null || text.isEmpty()) return 0;
+
+        String[] tokens = getNumbers(text);
+
+        int sum = 0;
+        for (String i : tokens) {
+            if (Integer.parseInt(i) < 0) {
+                throw new RuntimeException();
+            }
+            sum += Integer.parseInt(i);
+        }
+        return sum;
+    }
+
+    private String[] getNumbers(String text) {
+        String delimiter = ",|:";;
+        String numberString = text ;
+
+        Matcher m = Pattern.compile("//(.)\n(.*)").matcher(text);
```

Comment:
Pattern.compile static 메소드 호출 시 아래와 같이 new Pattern()을 반환하게 됩니다.

```java
    public static Pattern compile(String regex) {
        return new Pattern(regex, 0);
    }
```

Pattern 객체는 비용이 큰 객체이므로 위에 따로 정의하는건 어떨까요?

```java
    private static final Pattern CUSTOM_DELIMITER_PATTERN = Pattern.compile("//(.)\n(.*)");
```
```suggestion
    Matcher m = CUSTOM_DELIMITER_PATTERN.matcher(text);
```



```
String.matches는 내부적으로 Pattern 객체를 만들어 쓰는데 그 객체를 만들려면 정규 표현식으로 유한 상태 기계로 컴파일 하는 과정이 필요하다. 즉 비싼 객체다.

성능을 개선하려면 Pattern 객체를 만들어 재사용하는 것이 좋다.
```

- 관련 문서
https://github.com/keesun/study/blob/master/effective-java/item6.md